### PR TITLE
[clang] NFCI: Use `FileEntryRef` in `SourceManager::FileInfos`

### PR DESCRIPTION
**Pull request by jansvoboda11**

### CODE REVIEW

**Review by benlangmuir — COMMENTED** [review]
I'm skeptical that treating `FileEntryRef` as `FileEntry *` for the purposes of equality and hashing is a good idea.  To me, the obvious interpretation of `FileEntryRef ==` is that the path must be equal, and having it based on the dev/inode is surprising.  I know that's not something you're *introducing* in this PR, but I would prefer we remove that functionality rather than build more on top of it.

Here are some alternative ways we could handle this that I think are more principled:
* We could do `DenseMap<FileEntry *, FileEntryRef>` instead of the DenseSet, and similarly embed the FileEntryRef into the RHS type in cases with DenseMap
* Alternatively, we could add a wrapper type for FileEntryRef that continues to have the `FileEntry *` equality/hashing semantics, but where the name of that wrapper and/or the fact you need to unwrap it explicitly makes it obvious that the semantics are not "just" FileEntryRef.  If we can figure out a good name for this I think it would be a good solution.

**Comment by jansvoboda11** [comment]
I understand your concern, we should definitely look into this. However, I think the inode uniquing functionality is pretty fundamental to `FileEntry`, so I'm not sure it's reasonable for type named `FileEntryRef` to do anything else. I guess we could split `FileEntryRef` into two types that make it super clear whether they have the same uniquing semantics or not.

But, since we're only 5 more commits away from finally finishing the whole `FileEntry` to `FileEntryRef` transition, I think it would be more pragmatic to keep the status quo in this patch (it's been in place [since 2020](https://reviews.llvm.org/D92627)), finish the transition, and come up with the next steps after that's done. Note that none of the other 5 commits introduce any more usages of `FileEntryRef` in associative containers.

What do you think?

**Comment by benlangmuir** [comment]
For what it's worth, until recently there was almost no **use** of the implicit unique file semantics for FileEntryRef even though the DenseMapInfo had been available for a long time.  But okay, I see there are more uses now so this is not the right place to push back on that.

File: clang/unittests/Basic/FileEntryTest.cpp [hunk 1]
```diff
@@ -179,6 +179,46 @@ TEST(FileEntryTest, DenseMapInfo) {
     EXPECT_TRUE(Set.find(R1)->isSameRef(R1));
     EXPECT_TRUE(Set.find(R2)->isSameRef(R2));
   }
+
+  // Insert R1Also second and confirm R1 "wins" when looked up as FileEntry.
```
Comment (benlangmuir) [inline clang/unittests/Basic/FileEntryTest.cpp:183]: This comment looks like it's for the other test case.  R1Also is inserted first and it wins here.

**Review by benlangmuir — APPROVED** [review]
Comment in the test needs a fix, but otherwise LGTM

**Pull request merged**